containerd: add flag to allow host access

muntac opened the pull request:
<!--
Thanks for submitting a pull request to Concourse!

If you haven't already, feel free to [add yourself] as a contributor so that
you can add labels to your PR and re-trigger its builds if they fail.

Also check the [PR requirements] if you haven't already!
-->

[add yourself]: https://github.com/concourse/governance#individual-contributors
[PR requirements]: https://github.com/concourse/concourse/blob/master/CONTRIBUTING.md#pull-request-requirements

## Changes proposed by this PR
cc: @taylorsilva 

closes #6720

Chart PR: https://github.com/concourse/concourse-chart/pull/250
BOSH Release PR: https://github.com/concourse/concourse-bosh-release/pull/150
<!--
Summarize your changes as a checklist, leaving any unfinished work as unchecked
items. Please include reasoning and key decisions to help the reviewer
understand the changes.
-->

* [x] Block container from reaching host by default
* [x] Add a flag, `CONCOURSE_CONTAINERD_ALLOW_HOST_ACCESS`, to optionally allow container to reach host
* [x] Allow host access if DNS proxy is enabled

## Notes to reviewer

<!--
If needed, leave any special pointers for reviewing or testing your PR.
-->
1. Default behaviour

Set `CONCOURSE_CONTAINERD_DNS_PROXY_ENABLE: "false"` in `docker-compose.yml`.

```
docker-compose up -d
docker inspect concourse_default
# Note down the IPv4 address of concourse_worker_1
fly -t dev login -u test -p test
fly -t dev sp -p tw -c /topgun/task-waiting.yml
fly -t dev up -p tw
fly -t dev tj -j tw/simple-job
fly -t dev i -u http://localhost:8080/teams/main/pipelines/task-waiting/jobs/simple-job/builds/1
# On container
ping [IPv4 address without the CIDR range]
```

Ping should see no response.

2. Optionally allow host network access

Set `CONCOURSE_CONTAINERD_ALLOW_HOST_ACCESS: "true"`. Repeat the steps above. Ping should now work.

## Release Note

<!--
Your PR title will be directly included in the release notes when it ships in
the next Concourse release. It should briefly describe the PR in [imperative
mood]. Please refrain from adding prefixes like 'feature:', and don't include a
period at the end.

Within this section you may supply a list of extra information to include in
the release notes in addition to the pull request title.

Example title: Introduce new pipeline UI algorithm

Example notes:

* Reticulating splines is the new process Concourse uses to create the network
  of lines between jobs.
* Combines many short lines and curves into a network of splines.

If there are no additional notes necessary you may remove this entire section.
-->


* containerd: Previously containers were allowed to reach the host network by default. This is now prevented.
  * To allow traffic to the host network set `CONCOURSE_CONTAINERD_ALLOW_HOST_ACCESS` to true.
  * If DNS proxy is enabled host network access will be turned on implicitly.

muntac committed:
worker/runtime: block host access

Containers should not be able to reach the host's network by default.
Current implementation blocks the host's specific LAN IP.

Signed-off-by: Muntasir Chowdhury <mchowdhury@pivotal.io>
Co-authored-by: Taylor Silva <tsilva@pivotal.io>

muntac committed:
worker/runtime: Convert some tests to table tests

Co-authored-by: Taylor Silva <tsilva@pivotal.io>
Signed-off-by:    Muntasir Chowdhury <mchowdhury@pivotal.io>

taylorsilva committed:
worker: add flag to allow host access

Setting CONCOURSE_CONTAINERD_ALLOW_HOST_ACCESS to true will let
containers in the containerd runtime reach the worker host.

Signed-off-by: Taylor Silva <tsilva@pivotal.io>
Co-authored-by:    Muntasir Chowdhury <mchowdhury@pivotal.io>

taylorsilva committed:
containerd: allow host access when dns proxy enabled

DNS proxy requires that the container is able to access the host.  Setting
`CONCOURSE_CONTAINERD_DNS_PROXY_ENABLE` to "true" will now implicitly allow
host access.

Signed-off-by: Taylor Silva <tsilva@pivotal.io>
Co-authored-by: Muntasir Chowdhury <mchowdhury@pivotal.io>

aoldershaw commented on worker/runtime/cni_network.go:
```diff
@@ -303,6 +333,46 @@ func (n cniNetwork) generateResolvConfContents() ([]byte, error) {
+	if err != nil {
+		return err
+	}
+	n.ipt.AppendRule(filterTable, "INPUT", "-i", n.config.BridgeName, "-d", hostIp, "-j", "DROP")
```
Is there a reason for using `DROP` rather than `REJECT`? Seems like the main difference is that with `DROP`, trying to reach the host will timeout (rather than with `REJECT`, which will fail ASAP)

muntac replied:
`REJECT` explicitly responds to the attacker and lets them know that this IP exists. `DROP` prevents leaking that information.

aoldershaw replied:
That brings up another question, actually - does this rule only `DROP` packets destined for the IP of the `eth0` (or equivalent) interface? Will it prevent access to any IP that's accessible within the host's network?

e.g. I was investigating a customer issue where a VM's BOSH DNS server was inaccessible to containers, which was causing DNS issues. This DNS server had IP `169.254.0.2` (which wasn't the same as `eth0` on the host). Would this rule handle that case?

muntac replied:
Yup. It won't affect any other IPs. We tested this by bringing up a second worker and were able to ping that worker from the container. Guardian has the same behaviour.

Initially, we were wondering if we should try to stop that too but we concluded VM to VM network interactions in the network are up to the operator to determine through their firewall rules (they can technically prevent it through `CONCOURSE_CONTAINERD_RESTRICTED_NETWORK` as well).

muntac replied:
Not quite sure why the DNS issue came up in the customer issue. Possible it was their internal firewall rules or some nuance about Guardian (if they're using that runtime).

aoldershaw replied:
Sorry, should've clarified - it was a result of allow host access (i.e. with host access disabled, they couldn't access the internal DNS server, and could access it when enabled).

I guess the question I have is what's the reason for the different behaviour from Guardian? My interpretation of Guardian's `iptables` rules here are:
1. If packets are destined for the external network interface (eth0), then accept them
2. If packets are destined for any other interface, then reject if host access is disabled, and accept if enabled

...which seems like it should work more generally, as it seems local services can bind to other IP addresses (not sure how it works, but e.g. BOSH DNS binding to 169.254.0.2). Let me know if my interpretation is flawed, not 100% sure about how `iptables`/linux networking works evidently

muntac commented on worker/runtime/cni_network.go:
```diff
@@ -303,6 +333,46 @@ func (n cniNetwork) generateResolvConfContents() ([]byte, error) {
+	if err != nil {
+		return "", err
+	}
+	ethInterface := regexp.MustCompile("eth0")
```
Great catch! Will change this.

taylorsilva committed:
Somehow copy-paste guardian's iptable rule to block host access

I expanded the integration test to also check that the container can
still reach the internet. I then decided to see what would happen if I
used guardian's iptable rule that mystified us and found that the test
stil passed! Both methods work but the nice thing about guardian's is
that you don't need to know the LAN IP address of the host. It just
works (TM).

I have a theory about why this works now, not 100% sure how to validate
it though.

When a process in a container reaches out to the hosts's LAN IP, the
path the packet takes is:

proc in container -> concourse0 bridge -> host process (gdn, baggageclaim, etc.)

When the process in a container is reaching out to an internet IP
address the path it takes is:

proc in container -> eth0 interface and directly out

Maybe the kernel is smart enough to figure this out and skips the
container bridge? This would explain why adding the reject rule directly
on the filter's INPUT chain doesn't break everything.

Guardian also adds two other rules before the reject rule:

-A w--input -i eth0 -j ACCEPT
-A w--input -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT

These two never made sense to us. When we were investigating we
found no packets ever hit the first two rules, which makes sense. They
were part of a chain defined as:

-A INPUT -i w+ -j w--input

on the INPUT chain. This rule says, any packets going to an interface
that starts with "w" should go to the "w--input" chain. The first rule
in that chain:

-A w--input -i eth0 -j ACCEPT

makes no sense; packets on this chain are coming into "w+" interfaces,
it'll never match "eth0".

For the second rule, we get what it's meant to do, allow packets for
exisitng connections to pass through, we just don't know why you need it
here.

Assuming the above theory is correct, this single rule may be all we
need.

Signed-off-by: Taylor Silva <tsilva@pivotal.io>

taylorsilva commented:
@aoldershaw @muntac I made a change to the IP table rule being set. Read the commit message for an explanation: https://github.com/concourse/concourse/pull/7002/commits/a2629367cd006cacd1ea42f35928e50dc763cf7b

aoldershaw commented:
@taylorsilva I tried modifying the integration test to test the *opposite* case (i.e. with allow host access enabled, it should be able to reach the local server) - but it still fails. Like with guardian, do we need to add the opposite rule? (`-j ACCEPT` if allow host access is enabled)

taylorsilva commented:
@aoldershaw in that case the rule is not written at all and the `INPUT` chain's default policy of `ACCEPT` would allow the packet through.

I'm trying to test this right now but I'm getting a timeout issue, which shouldn't happen since I changed it to return a response upon rejection 😡  Will look into this more on Monday.

taylorsilva commented:
I think I see why it failed. The iptables rules are not flushed between tests, so it's possible for the env to become polluted and subsequent tests to fail. Will update the tests with a fix.

taylorsilva committed:
Add test to verify host access is allowed when enabled

Fixed the Before and After test suite hooks. They weren't actually
running, now they are. Tests that configured a custom backend should
have also stopped and cleaned up what was created in the BeforeTest()
but they didn't which is why the AllowHostAccess test was failing.

Signed-off-by: Taylor Silva <tsilva@pivotal.io>

taylorsilva commented:
I ended up NOT being wrong, but in a different way 😅 

Tests weren't being cleaned up properly and for tests that used a custom backend they didn't cleanup stuff made in the `BeforeTest()` before making their own backend. They were starting in a dirty environment.

Everything should work as expected now!

aoldershaw reviewed (approved):
LGTM. I was a bit concerned the `REJECT` rule to the bridge interface would make https://github.com/concourse/rfcs/pull/84 impossible (without some additional service manager component running on the worker to selectively grant containers access to others). However, it doesn't seem to - I disabled the DNS proxy, spun up two tasks, and was able to communicate between them using `nc`. Is this because it's on the `INPUT` chain rather than the `FORWARD` chain, and packets from one container aren't "entering the system" from the outside?

taylorsilva commented:
> Is this because it's on the INPUT chain rather than the FORWARD chain, and packets from one container aren't "entering the system" from the outside?

This area is still kinda grey for me. Haven't figured out a good way to verify behaviour around interfaces + iptable rules, but the last part you said, the packet ""entering the system" from the outside", is the key part I think and is what causes it to not hit the REJECT rule in this case.
